### ISSUE TITLE
Actionlib Handle assertion killed program

### ISSUE THREAD

**Issue by pbeeson** [ros/actionlib#67]
I have a pretty intensive robotic application that we are running through reliability tests before software goes into production in June.  After about a week of running fine, one our actionlib clients died due to a low-level assertion in the Handle class.

```
affordance_template_server_node: /home/pbeeson/ros/vehtek/src/actionlib/include/actionlib/managed_list.h:171: bool actionlib::ManagedList<T>::Handle::operator==(const actionlib::ManagedList<T>::Handle&) const [with T = boost::shared_ptr<actionlib::CommStateMachine<affordance_template_msgs::PlanAction_<std::allocator<void> > > >]: Assertion `valid_' failed.
```

**Comment by pbeeson** [comment]
Could it be that on line 182 of managed_list.h, that the `valid_(true)` (the only place valid_ is ever set to true) should come before the initialization of `it_` and `handle_tracker_`?

**Comment by pbeeson** [comment]
So it looks like the Handle reset() function is being continuously called as the ClientGoalHandle instance is created/destroyed constantly.  Because this exact object that is being examined is a ptr to the CommStateMachine, I"m wondering if this isn't another race condition in the GoalManager class like Issue #66

**Comment by pbeeson** [comment]
My current "fix" for this is to go through the client code and replace all asserts with ROS_ERROR messages, and return default values where appropriate.

[Temp fix](https://github.com/pbeeson/actionlib/commit/d761eadb7bb2fbbe3bd23e41195cb106135dcff8)

**Comment by pbeeson** [comment]
Anyone have thoughts on this?

**Comment by pbeeson** [comment]
After 5 days of running, I haven't encountered the error message I replaced the assertions with, but I am seeing this:
`Got a callback on a goalHandle that we're not tracking.  This is an internal SimpleActionClient/ActionClient bug.  This could also be a GoalID collision`

My feeling is that this symptom is from the same underlying problem, in that goals can be deleted or partially deleted in the client from a Done/Finished message before some later feedback messages get processed about the now missing goal in the actionlib client.

Thankfully, this doesn't seem to really bring down my code.

**Comment by mikaelarguedas** [comment]
addressed in #83. Note the code still needs to be compiled in Release/RelWithDebInfo build type for the asserts to be removed
